File: leetcode/1704. Determine if String Halves Are Alike/main.go
// Solved by @kitanoyoru

package main

func checkIsVowel(ch rune) bool {
	var vowels = map[rune]bool{
		'a': true,
		'e': true,
		'i': true,
		'o': true,
		'u': true,
		'A': true,
		'E': true,
		'I': true,
		'O': true,
		'U': true,
	}

	return vowels[ch]
}

func halvesAreAlike(s string) bool {
	c1, c2 := 0, 0
	mid := len(s) / 2

	for i := 0; i < mid; i++ {
		if checkIsVowel(rune(s[i])) {
			c1++
		}
	}
	for i := mid; i < len(s); i++ {
		if checkIsVowel(rune(s[i])) {
			c2++
		}
	}

	return c1 == c2
}

File: leetcode/1704. Determine if String Halves Are Alike/main.rs
impl Solution {
    pub fn halves_are_alike(s: String) -> bool {
        assert!(s.len() % 2 == 0);

        let mid = s.len() / 2;
        let (first, second) = s.split_at(mid);

        Self::count_vowels(first) == Self::count_vowels(second)
    }

    fn count_vowels(s: &str) -> usize {
        let vowels: [char; 5] = ['a', 'e', 'i', 'o', 'u'];

        s.chars()
            .filter(|c| {
                vowels.contains(&c.to_ascii_lowercase()) || vowels.contains(&c.to_ascii_uppercase())
            })
            .count()
    }
}
